integrate: use corrected velocity grads with Navier-Stokes

GonzaloSaez opened the pull request:
Hi everyone,

It seems that when analyzing the pressure-dilatation terms  of the default TGV vortex [1] using the integrate plugin as

`int-dissipationPressure = p*(grad_u_x + grad_v_y + grad_w_z)`

, the results are very far from the expected ones (two-order of magnitudes higher), although other quantities such as kinetic energy dissipation are very close to those shown in the literature. I also observed these issues with my in-house code (which uses a similar approach to compute the gradients in the postprocessing stages). Therefore, I tried to fix the issue by relying on the corrected gradients of the velocity at solution points, instead of the gradients computed using the nodal solution basis. When using these corrected gradients, the results are much closer to those obtained in the literature. 

I suppose that this is not the cleanest implementation (only compatible with the `navier-stokes` solver) and it will harm the performance (extra copy operations) and also increase the spatial complexity of the implementation (due to the need of storing a copy of `_vect_upts`). Anyway, I'll leave this implementation in the PR in case anyone needs to postprocess this pressure dilation term with PyFR.

[1] DeBonis, J. (2013), “Solutions of the Taylor-Green vortex problem using high-resolution explicit finite
difference methods”, In 51st AIAA Aerospace Sciences Meeting including the New Horizons Forum and Aerospace Exposition, p. 382.

FreddieWitherden commented:
So I've been thinking about this for a while now.  The final solution I came up with (although never implemented) was to add a new method alongside rhs() which simply does enough work to compute the gradients and then stops.  Then, if a plugin needs gradients, this method can be called.

The advantages here are twofold: there is no performance hit when gradients are not needed by a plugin or when the plugin is only run periodically; and there is no time lag between the gradients and the solution.  (The gradients your code picks up are typically from the final RK stage which is usually working on a very different solution to the final solution generated by the RK scheme.)

Some extra work would also be required in each of the solvers to state how to convert conserved gradients into primitive gradients, although this should not be too difficult.

GonzaloSaez commented:
Hi Freedie,

Thank you for your answer. I understand your concerns regarding the lag between the gradients and the conservative variables values at the end of the RK step (although for postprocessing purposes and if the time step is small enough the differences might not be very important). To better implement this PR, I suppose that one might want to add the following function

```python
class BaseAdvectionDiffusionSystem(BaseAdvectionSystem):
...
def compute_grads(self, t, uinbank, foutbank):
        runall = self.backend.runall
        q1, q2 = self._queues
        kernels = self._kernels

        self._bc_inters.prepare(t)

        self.eles_scal_upts_inb.active = uinbank
        self.eles_scal_upts_outb.active = foutbank

        if ('eles', 'copy_soln') in kernels:
            q1.enqueue(kernels['eles', 'copy_soln'])
        if ('iint', 'copy_fpts') in kernels:
            q1.enqueue(kernels['iint', 'copy_fpts'])
        q1.enqueue(kernels['iint', 'con_u'])
        q1.enqueue(kernels['bcint', 'con_u'], t=t)
        if ('eles', 'shocksensor') in kernels:
            q1.enqueue(kernels['eles', 'shocksensor'])
            q1.enqueue(kernels['mpiint', 'artvisc_fpts_pack'])
        q1.enqueue(kernels['eles', 'tgradpcoru_upts'])
        q2.enqueue(kernels['mpiint', 'scal_fpts_send'])
        q2.enqueue(kernels['mpiint', 'scal_fpts_recv'])
        q2.enqueue(kernels['mpiint', 'scal_fpts_unpack'])

        runall([q1, q2])

        q1.enqueue(kernels['mpiint', 'con_u'])
        q1.enqueue(kernels['eles', 'tgradcoru_upts'])
        q1.enqueue(kernels['eles', 'gradcoru_upts'])

        runall([q1])
````

which should be called by the considered plugin using appropriate banks. Moreover we should implement the function `grad_con_to_pri` to `BaseFluidElements` to transform from conservative to primitive gradients. I'll try to take a look to see if this is feasible without a lot of work

FreddieWitherden commented:
> I understand your concerns regarding the lag between the gradients and the conservative variables values at the end of the RK step

A larger issue is that the solutions used by the individual stages of an RK scheme are typically of much lower order than the final solution generated by the scheme.  In the RK literature this is known as the 'stage order.'  The only good solution is the final solution yielded by the scheme; all of the intermediate solutions should be regarded as suspect.  Only when they are combined (right at the end) do you obtain a useful solution.  Hence the issue with taking the gradients picked up from one of these intermediate solutions.

GonzaloSaez commented:
Hi @FreddieWitherden , let me know if this new implementation works for you. There might some more details to be discussed (the implementation is only compatible with the solver `navier-stokes` although it could be easily extended to `ac-navier-stokes`). I'm also not sure if the gradient computation method is compatible with other than `std` integrators (see the logic used to chose `regidx` in `integrate.__call__`)

FreddieWitherden commented:
> Hi @FreddieWitherden , let me know if this new implementation works for you. There might some more details to be discussed (the implementation is only compatible with the solver `navier-stokes` although it could be easily extended to `ac-navier-stokes`). I'm also not sure if the gradient computation method is compatible with other than `std` integrators (see the logic used to chose `regidx` in `integrate.__call__`)

So I think we want a mechanism for plugins to get gradients generically like they can get the solution with `intg.soln`.  For example `intg.grad_soln` which they can then convert to primitive form by calling the relevant method.  This attribute can also adopt a similar caching mechanism to what we use with `intg.soln`.

FreddieWitherden commented on pyfr/solvers/baseadvecdiff/system.py:
```diff
@@ -63,3 +63,38 @@ def rhs(self, t, uinbank, foutbank):
         q1.enqueue(kernels['eles', 'negdivconf'], t=t)
         runall([q1])
+
+    def compute_grads(self, t, uinbank, foutbank):
```
No need for the outbank argument.

FreddieWitherden commented on pyfr/solvers/baseadvecdiff/system.py:
```diff
@@ -63,3 +63,38 @@ def rhs(self, t, uinbank, foutbank):
+        self._bc_inters.prepare(t)
+
+        self.eles_scal_upts_inb.active = uinbank
+        self.eles_scal_upts_outb.active = foutbank
```
No need for this.

FreddieWitherden commented on pyfr/solvers/baseadvecdiff/system.py:
```diff
@@ -63,3 +63,38 @@ def rhs(self, t, uinbank, foutbank):
+        q1.enqueue(kernels['eles', 'disu'])
+        q1.enqueue(kernels['mpiint', 'scal_fpts_pack'])
+        runall([q1])
+        if ('eles', 'copy_soln') in kernels:
```
No need for this.

FreddieWitherden commented on pyfr/solvers/baseadvecdiff/system.py:
```diff
@@ -63,3 +63,38 @@ def rhs(self, t, uinbank, foutbank):
+            q1.enqueue(kernels['iint', 'copy_fpts'])
+        q1.enqueue(kernels['iint', 'con_u'])
+        q1.enqueue(kernels['bcint', 'con_u'], t=t)
+        if ('eles', 'shocksensor') in kernels:
```
No need for this.

FreddieWitherden commented on pyfr/solvers/baseadvecdiff/system.py:
```diff
@@ -63,3 +63,38 @@ def rhs(self, t, uinbank, foutbank):
+
+        q1.enqueue(kernels['mpiint', 'con_u'])
+        q1.enqueue(kernels['eles', 'tgradcoru_upts'])
+        q1.enqueue(kernels['eles', 'gradcoru_upts'])
```
This has changed with the most recent version of PyFR.

FreddieWitherden commented on pyfr/solvers/navstokes/elements.py:
```diff
@@ -8,6 +8,32 @@ class NavierStokesElements(BaseFluidElements, BaseAdvectionDiffusionElements):
+        vs = [rhov/rho for rhov in cons[1:-1]]
+
+        # velocity gradients
+        grad_v = [(grad_cons[i + 1] - vel*grad_rho)/rho for i, vel in enumerate(vs)]
```
zip rather than enumerate

FreddieWitherden commented on pyfr/solvers/navstokes/elements.py:
```diff
@@ -8,6 +8,32 @@ class NavierStokesElements(BaseFluidElements, BaseAdvectionDiffusionElements):
+        # pressure gradient
+        gamma = cfg.getfloat('constants', 'gamma')
+        vsqr = sum(v*v for v in vs)
+        grad_p = grad_E.copy()
```
I think this can be cleaned up.

GonzaloSaez commented:
> > Hi @FreddieWitherden , let me know if this new implementation works for you. There might some more details to be discussed (the implementation is only compatible with the solver `navier-stokes` although it could be easily extended to `ac-navier-stokes`). I'm also not sure if the gradient computation method is compatible with other than `std` integrators (see the logic used to chose `regidx` in `integrate.__call__`)
> 
> So I think we want a mechanism for plugins to get gradients generically like they can get the solution with `intg.soln`. For example `intg.grad_soln` which they can then convert to primitive form by calling the relevant method. This attribute can also adopt a similar caching mechanism to what we use with `intg.soln`.

I have implemented the `intg.grad_soln` method to ensure that the gradients are only computed once if multiple plugins need to access the corrected gradient data. I suppose that the way in which I store `vect_upts_inb` might not be ideal, so please let me know if you think that there is another way of storing the banks of the gradients.

I also fixed all your comments (let me know if not), rebased the branch with develop and modified the `compute_grads` function to take into account linear and curved elements.

PD: the `intg.grad_soln` needs to also be implemented with other than `std` integrators, I'll do that after your review of the current implementation.

FreddieWitherden commented on pyfr/plugins/integrate.py:
```diff
@@ -21,10 +21,10 @@ def __init__(self, intg, cfgsect, suffix=None):
 
         # Underlying system
-        system = intg.system
+        self.system = intg.system
```
Think these changes can be reverted.

GonzaloSaez replied:
True, I'll revert them

FreddieWitherden commented on pyfr/plugins/integrate.py:
```diff
@@ -130,7 +130,7 @@ def _eval_exprs(self, intg):
 
         # Iterate over each element type in the simulation
-        for i, (soln, eleinfo) in enumerate(zip(intg.soln, self.eleinfo)):
+        for i, (soln, grad_soln, eleinfo) in enumerate(zip(intg.soln, intg.grad_soln, self.eleinfo)):
```
I'd only pull the property inside of the gradpnames block; avoids computing the gradient if it is not needed.

GonzaloSaez replied:
Yes you're right

FreddieWitherden commented on pyfr/solvers/acnavstokes/elements.py:
```diff
@@ -6,6 +6,13 @@
 
 class ACNavierStokesElements(BaseACFluidElements,
                              BaseAdvectionDiffusionElements):
+    gradconvarmap = {k : [f'grad_{var}' for var in v] for k, v in BaseACFluidElements.convarmap.items()}
```
Unsure if we need these maps; they can always be synthesised from convarmap and privarmap.

GonzaloSaez replied:
True, I'll then check if the gradient to be computed is the gradient of a primitive variable or not

FreddieWitherden commented on pyfr/solvers/base/elements.py:
```diff
@@ -212,6 +212,10 @@ def set_backend(self, backend, nscalupts, nonce, linoff):
 
+        # Allocate and bank the storage required to store the solution grads
+        # which will be used by post-processing plugins
+        self.vect_upts_inb = backend.matrix_bank([self._vect_upts])
```
No need to bank this.

GonzaloSaez replied:
Do you then suggest to use `self.vect_upts_ib = self._vect_upts`? I'm sorry if these questions are a bit silly, I'm not fully used to the way in which data is accessed in PyFR

FreddieWitherden replied:
Yes; this is on account of the fact that there is only one self._vect_upts array (unlike scalar arrays where due to the RK schemes there can be a bunch).

FreddieWitherden commented:
This is shaping up nicely.

I think the next question we want to ask is if the gradient property should fall back to computing gradients the direct way (by differentiating the polynomials) if the system does not provide a proper method.  This would enable everything to work for Euler and friends (although the gradients are, of course, not as good).

GonzaloSaez commented:
> This is shaping up nicely.
> 
> I think the next question we want to ask is if the gradient property should fall back to computing gradients the direct way (by differentiating the polynomials) if the system does not provide a proper method. This would enable everything to work for Euler and friends (although the gradients are, of course, not as good).

I think that this could be done by checking if `system` inherits of `BaseAdvecDiff`. In that case, we can rely on the corrected LDG gradients, else we rely on the old implementation.

FreddieWitherden commented:
> > This is shaping up nicely.
> > I think the next question we want to ask is if the gradient property should fall back to computing gradients the direct way (by differentiating the polynomials) if the system does not provide a proper method. This would enable everything to work for Euler and friends (although the gradients are, of course, not as good).
> 
> I think that this could be done by checking if `system` inherits of `BaseAdvecDiff`. In that case, we can rely on the corrected LDG gradients, else we rely on the old implementation.

It might be a tiny bit more subtle as to do the untransformation some metric terms are needed.  And I believe by the time the function is called that these metric terms are no longer readily available.  (This is why things like the integrator plugin grab a reference to them when they're initialised.)  Doing this all the time might be okay, although it could also be a waste of memory.

GonzaloSaez commented:
> > > This is shaping up nicely.
> > > I think the next question we want to ask is if the gradient property should fall back to computing gradients the direct way (by differentiating the polynomials) if the system does not provide a proper method. This would enable everything to work for Euler and friends (although the gradients are, of course, not as good).
> > 
> > 
> > I think that this could be done by checking if `system` inherits of `BaseAdvecDiff`. In that case, we can rely on the corrected LDG gradients, else we rely on the old implementation.
> 
> It might be a tiny bit more subtle as to do the untransformation some metric terms are needed. And I believe by the time the function is called that these metric terms are no longer readily available. (This is why things like the integrator plugin grab a reference to them when they're initialised.) Doing this all the time might be okay, although it could also be a waste of memory.

Yes that's true, maybe this can be discussed in other PR? I have already introduced your suggestions in the implementation, although I still don't know how to access the data of `self._vect_upts` without a bank.

FreddieWitherden commented on pyfr/solvers/base/system.py:
```diff
@@ -42,6 +42,9 @@ def __init__(self, backend, rallocs, mesh, initsoln, nregs, cfg):
         self.ele_shapes = [(e.nupts, e.nvars, e.neles) for e in eles]
 
+        # Get the gradient banks
+        self.ele_grad_banks = list(eles.vect_upts_inb)
```
I do not think we need this, we can just call .get() on inb.

GonzaloSaez replied:
Done

FreddieWitherden commented on pyfr/integrators/std/base.py:
```diff
@@ -47,6 +47,15 @@ def soln(self):
+        # If we do not have the solution gradients cached then compute and fetch them
+        if not self._curr_grad_soln:
+            self.system.compute_grads(self.tcurr, self._idxcurr)
+            self._curr_grad_soln = self.system.ele_vect_upts()
```
Can just fetch directly.

GonzaloSaez replied:
Done

FreddieWitherden commented on pyfr/plugins/integrate.py:
```diff
@@ -152,12 +153,31 @@ def _eval_exprs(self, intg):
+                    # rely on the LDG corrected gradients to compute
+                    # the primitive gradients, instead of
+                    # evaluating them from the solution nodal basis
+                    if pname in self.elementscls.privarmap[self.ndims] \
```
Ideally this code should not depend on the system but should be abstracted elsewhere.

GonzaloSaez replied:
Abstracted in the `system` class maybe?

Should we also use these gradients in `fluidforce.py`?

FreddieWitherden replied:
Fluid force is a bit of a special case as it needs gradients on the surfaces and only for a small subset of the elements.  So, for the moment I'd be inclined to keep fluid force the same.  The time averaging plugin, however, also uses gradients so it should be switched over as appropriate.

GonzaloSaez replied:
I see, it seems reasonable to avoid the computation of corrected gradients for `fluidforce`. For the time averaging plugin, we could reuse the same logic as in `integrate` to use or not the  corrected primitive gradients although we need to find a sensible way of implementing this logic. Maybe we could add a function to `system` to decide whether a given gradient can be computed using the LDG method or not?

FreddieWitherden replied:
Yeah, system or integrator seems like the right place.  This way plugins can just ask for gradients and do not need to concern themselves about where they are coming from.

GonzaloSaez replied:
I've come up with this implementation to compute the gradients in `intg`. `compute_grads` raises a `NotImplementedError` if the `system` does not explictly compute the gradients. In that case, the gradients are computed without LDG correction. However, too handle both cases we need to add a lot of parameters to the function (metric terms, gradient operator, element type index, etc.). I suppose that that's a bit ugly, do you have a better idea?

```python
def grad_pvar(self, pname, gradop, rcpjact, ele_index, eset):
        if pname not in self.system.elementscls.privarmap[self.system.ndims]:
            raise ValueError(f'Gradient of non-primitive variable {pname} can not be computed')
        # Subset and transpose the solution
        soln = self.soln[ele_index]
        soln = soln[..., eset].swapaxes(0, 1)
        try:
            # Fetch the corrected gradients only if needed
            grad_soln = self.grad_soln[ele_index]

            # Subset and transpose the solution gradients
            grad_soln = grad_soln[..., eset].swapaxes(0, 2).swapaxes(1, 2)

            # Transform from conservative to primitive gradients
            pgrads = self.system.elementscls.grad_con_to_pri(soln, grad_soln, self.cfg)
            idx = self.system.elementscls.privarmap[self.system.ndims].index(pname)
            return pgrads[idx]
        except NotImplementedError:
            # If the considered solver does not explictly compute gradients
            # of primitive variables, then compute the gradient from
            # the solution nodal basis

            # Number of solution points
            nupts = gradop.shape[1]

            # Convert from conservative to primitive variables
            psolns = self.system.elementscls.con_to_pri(soln, self.cfg)

            psoln = psolns[psolns.index(pname)]
            # Compute the transformed gradient
            tgradpn = gradop @ psoln
            tgradpn = tgradpn.reshape(self.ndims, nupts, -1)

            # Untransform this to get the physical gradient
            gradpn = np.einsum('ijkl,jkl->ikl', rcpjact, tgradpn)
            gradpn = gradpn.reshape(self.ndims, nupts, -1)
            return gradpn
```

FreddieWitherden replied:
I agree this is quite a bit of work.  What are your thoughts on only allowing gradients for advection-diffusion derived systems?  This would greatly simplify the code and enable us to always provide `good' gradients.  I do not think many people make use of gradients in Euler simulations.

GonzaloSaez replied:
I suppose that gradients might be important in Euler to tweak shock sensors, but I think that those values are used in a post-processing stage. You know much better this code than me so I can not really tell you if it is worth the effort. I agree that the implementation is a bit ugly but all plugins have direct access to those parameters so I really don't know, I'll let you decide!

FreddieWitherden replied:
My inclination is to drop support.  If there are really important use cases we can consider adding it back in later.

GonzaloSaez replied:
Let me know if the last commit works for you. The function `intg.grad_pvars` computes all the wanted gradients at once. I also erased the previous implementation from `integrate`. Maybe we should the same for `tavg`.

FreddieWitherden commented on pyfr/solvers/navstokes/elements.py:
```diff
@@ -8,6 +8,26 @@ class NavierStokesElements(BaseFluidElements, BaseAdvectionDiffusionElements):
+    @staticmethod
+    def grad_con_to_pri(cons, grad_cons, cfg):
+        rho = cons[0]
+        grad_rho = grad_cons[0]
```
grad_rho, *grad_uvw, grad_E = grad_cons

FreddieWitherden replied:
Just following up here.

FreddieWitherden commented on pyfr/solvers/navstokes/elements.py:
```diff
@@ -8,6 +8,26 @@ class NavierStokesElements(BaseFluidElements, BaseAdvectionDiffusionElements):
+        # Pressure gradient
+        gamma = cfg.getfloat('constants', 'gamma')
+        grad_p = grad_E - 0.5*sum(v*v for v in vs)*grad_rho
+        for vel, grad_vel in zip(vs, grad_v):
```
Can avoid multiply through by rho by using conservative grads

FreddieWitherden replied:
I think np.dot here with the right set of arguments could be used here (so it performs a bunch of smaller dot products for us).

FreddieWitherden commented on pyfr/solvers/base/system.py:
```diff
@@ -186,6 +187,9 @@ def _gen_kernels(self, eles, iint, mpiint, bcint):
         pass
 
+    def compute_grads(self, t, uinbank):
+        pass
```
Maybe have this raise a not implemented error and have the integrator catch this exception to decide when to fallback.

FreddieWitherden commented on pyfr/integrators/base.py:
```diff
@@ -181,3 +185,33 @@ def _add(self, *args):
         # Bind and run the axnpby kernels
         self._queue.enqueue_and_run(axnpby, *args[::2])
+
+    def grad_pvars(self, pnames, ele_index, eset):
```
I would just have it return everything (no pnames/ele_index, or eset arg).

GonzaloSaez replied:
Done,  I had move the function to `BaseIntegrator` to ensure compatibility with other than `std` integrators

FreddieWitherden commented on pyfr/integrators/base.py:
```diff
@@ -130,6 +134,38 @@ def cfgmeta(self):
         else:
             return {'config': cfg, 'config-0': cfg}
 
+    def grad_pvars(self, pnames):
```
I'd also remove the pnames bit, too.  Let each plugin handle messing with the data as it sees fit.

GonzaloSaez replied:
Done

FreddieWitherden commented on pyfr/integrators/base.py:
```diff
@@ -130,6 +134,32 @@ def cfgmeta(self):
+        try:
+            # Fetch the corrected gradients only if needed
+            grad_soln = self.grad_soln
+        except NotImplementedError:
```
No need to try/catch; instead we can stick a better message in when we raise the NotImplementedError.

GonzaloSaez replied:
Do you prefer raising an error in the NotImplementedError then?

FreddieWitherden replied:
Yes, I think that is the most logical place.

FreddieWitherden commented on pyfr/integrators/base.py:
```diff
@@ -130,6 +134,32 @@ def cfgmeta(self):
+            soln = soln.swapaxes(0, 1)
+
+            # Subset and transpose the solution gradients
+            grad_soln = grad_soln.swapaxes(0, 2).swapaxes(1, 2)
```
I think this might be equivalent to a rollaxis?

GonzaloSaez replied:
I'll have a look at it, I'm not very familiar with `numpy` tbh

FreddieWitherden replied:
```
In [1]: import numpy as np

In [2]: x = np.random.randn(10, 20, 30)

In [3]: np.allclose(x.swapaxes(0, 2).swapaxes(1, 2), np.rollaxis(x, 2))
Out[3]: True
```

However, this code should probably be moved out to the plugins themselves (integrate and tavg).

GonzaloSaez replied:
Added the `rollaxis` call and also modified `tavg` to use these new gradients

FreddieWitherden commented on pyfr/integrators/base.py:
```diff
@@ -130,6 +134,32 @@ def cfgmeta(self):
+            pgrads = self.system.elementscls.grad_con_to_pri(soln, grad_soln, self.cfg)
+
+            # Store the gradients
+            for grad in pgrads:
```
.extend()

FreddieWitherden commented on pyfr/solvers/base/elements.py:
```diff
@@ -212,6 +212,10 @@ def set_backend(self, backend, nscalupts, nonce, linoff):
 
+        # Allocate and bank the storage required to store the solution grads
+        # which will be used by post-processing plugins
+        self.vect_upts_inb = self._vect_upts
```
Don't think we want to call this an `inb` (which is short for input bank, but it is not a bank per se and isn't used as an input).

GonzaloSaez replied:
So maybe access directly `self._vect_upts` and rename `_vect_upts` to `vect_upts`?

GonzaloSaez replied:
Removed this `_vect_upts_inb`, the implementation now accesses `_vect_upts` directly

FreddieWitherden commented on pyfr/solvers/base/system.py:
```diff
@@ -186,6 +187,9 @@ def _gen_kernels(self, eles, iint, mpiint, bcint):
         pass
 
+    def compute_grads(self, t, uinbank):
+        raise NotImplementedError()
```
We can improve the error message here to indicate the system in question.

GonzaloSaez replied:
Done

FreddieWitherden commented on pyfr/plugins/integrate.py:
```diff
@@ -8,6 +8,7 @@
 from pyfr.nputil import npeval
 from pyfr.plugins.base import BasePlugin, init_csv
 from pyfr.quadrules import get_quadrule
+from pyfr.solvers.baseadvecdiff import BaseAdvectionDiffusionSystem
```
Not sure if we need this import.

GonzaloSaez replied:
Removed

FreddieWitherden commented on pyfr/solvers/acnavstokes/elements.py:
```diff
@@ -6,6 +6,10 @@
                              BaseAdvectionDiffusionElements):
+    @staticmethod
+    def grad_con_to_pri(cons, grad_cons, cfg):
+        return [grad for grad in grad_cons]
```
Can we just return grad_cons or list(grad_cons) here?

GonzaloSaez replied:
Done

FreddieWitherden commented on pyfr/integrators/base.py:
```diff
@@ -130,6 +134,26 @@ def cfgmeta(self):
         else:
             return {'config': cfg, 'config-0': cfg}
 
+    def grad_pvars(self):
```
I'd move this functionality down into each plugin.  This way that can fetch the gradients, subset them, and then only convert the subsetted parts to primitive.

GonzaloSaez replied:
Done for `integrate`, let me know if its ok for you and I'll move it also to `tavg`

FreddieWitherden commented on pyfr/integrators/base.py:
```diff
@@ -1,6 +1,7 @@
 # -*- coding: utf-8 -*-
 
 from collections import deque
+from collections import Iterable
```
Not sure if we need this import.

GonzaloSaez replied:
Removed

FreddieWitherden commented on pyfr/integrators/base.py:
```diff
@@ -130,7 +133,6 @@ def cfgmeta(self):
         else:
             return {'config': cfg, 'config-0': cfg}
 
```
Think this newline should be there.

GonzaloSaez replied:
Fixed

FreddieWitherden commented on pyfr/plugins/integrate.py:
```diff
@@ -172,6 +165,29 @@ def _eval_exprs(self, intg):
+        grads_eles = []
+        # Iterate over each element type in the simulation
+        for i, (soln, eleinfo) in enumerate(zip(intg.soln, self.eleinfo)):
+            plocs, wts, eset, emask = eleinfo
```
eset, emask = eleinfo[2:] as I don't think we need the plocs of wts here.

GonzaloSaez replied:
Fixed

FreddieWitherden commented on pyfr/plugins/tavg.py:
```diff
@@ -118,23 +122,12 @@ def _eval_acc_exprs(self, intg):
-                    gradpn = gradpn.reshape(self.ndims, nupts, -1)
-
+                    idx = self.elementscls.privarmap[self.ndims].index(pname)
+                    gradpn = gradps[idx][..., rgn]
```
I think we've already subsetted the gradients here?

GonzaloSaez replied:
I had to update the implementation in `tavg`. Now it is more consistent with `integrate`, let me know if you think it is OK

FreddieWitherden commented on pyfr/plugins/integrate.py:
```diff
@@ -172,6 +165,29 @@ def _eval_exprs(self, intg):
+
+        grads_eles = []
+        # Iterate over each element type in the simulation
+        for i, (soln, eleinfo) in enumerate(zip(intg.soln, self.eleinfo)):
```
It's also worth noting that the integrate and time averaging plugins use different region/subsetting logic and so I am unsure how much code they can share.  (Integrate can sometimes select specific points from specific elements, whereas tavg will also work in terms of full elements.)

GonzaloSaez replied:
Yes, the last version includes `_grad_pvars` with the logic of `tavg`

FreddieWitherden replied:
Okay, would you be able to test if everything still works as expected for regions which 'cut' through an element.  (Here PyFR should just include the quadrature points which are inside the region for boundary elements; which is a reasonable-ish assumption.)

GonzaloSaez replied:
There was a bug when meshes had more than one type of element due to the fact that `grad_soln` was being reused when calling `np.rollaxis`. This has been solved in the last commit.

I have tried the inc-cylinder-2d test case and it seems that `tavg` works okeysh even when using an arbitrary region box, I don't know if you have a better validation testcase for the `tavg` plugin. I'll leave attached a picture of the mean `u` velocity in the cylinder test case (the averaging starts from `t=0` so this result is not perfect but it should validate a bit the implementation)

![image](https://user-images.githubusercontent.com/11050889/107158625-bb610380-698b-11eb-8ac7-8caa0d8e8268.png)

GonzaloSaez replied:
I'll try tomorrow to output the mean velocity gradients

FreddieWitherden replied:
The other bug I watch out for is when a region only has a single type of element (but the partition itself has two or more element types).

FreddieWitherden commented on pyfr/plugins/tavg.py:
```diff
@@ -118,23 +122,12 @@ def _eval_acc_exprs(self, intg):
-                    gradpn = np.einsum('ijkl,jkl->ikl', rcpjact, tgradpn)
-                    gradpn = gradpn.reshape(self.ndims, nupts, -1)
-
+                    idx = self.elementscls.privarmap[self.ndims].index(pname)
```
We might want to pre-compute these indices and store them in self._gradpnames (suitably renamed, of course).  This should enable us to simplify these two loops.

GonzaloSaez replied:
`self_gradpidxs`?

FreddieWitherden replied:
Something like that, where the variable itself is a list of tuples (name, idx).

GonzaloSaez replied:
Done

FreddieWitherden commented on pyfr/plugins/integrate.py:
```diff
@@ -106,6 +106,10 @@ def _init_gradients(self, intg, rinfo):
 
+        privarmap = self.elementscls.privarmap[self.ndims]
+        self._gradpinfo = [(pname, privarmap.index(pname)) 
+                            for pname in self._gradpnames]
```
Can use gradpnames directly here (and we no longer need to store self._gradpnames, since all of the relevant info is now in self._gradpinfo).

GonzaloSaez replied:
Done

FreddieWitherden commented on pyfr/plugins/integrate.py:
```diff
@@ -143,23 +151,11 @@ def _eval_exprs(self, intg):
-
+                gradps = grads_eles[i]
+                for pname, idx in self._gradpinfo:
+                    gradpn = gradps[idx]
```
```python
                for pname, idx in self._gradpinfo:
                    for dim, grad in zip('xyz', grads_eles[i][idx]):
                        subs[f'grad_{pname}_{dim}'] = grad
```

GonzaloSaez replied:
Done

FreddieWitherden commented on pyfr/plugins/integrate.py:
```diff
@@ -143,23 +151,11 @@ def _eval_exprs(self, intg):
 
-            # Compute any required gradients
+            # Prepare any required gradient
             if self._gradpnames:
```
I think we can remove this check.

GonzaloSaez replied:
I think that this is needed in order to compute all gradients before looping over the regions and elements. This could maybe be avoided if we passed the region information to `_grad_pvars`

FreddieWitherden commented on pyfr/plugins/tavg.py:
```diff
@@ -118,23 +126,11 @@ def _eval_acc_exprs(self, intg):
-                    gradpn = np.einsum('ijkl,jkl->ikl', rcpjact, tgradpn)
-                    gradpn = gradpn.reshape(self.ndims, nupts, -1)
-
+                gradps = grads_eles[j]
```
See proposed simplification for the integrate plugin.

GonzaloSaez replied:
Done

FreddieWitherden commented on pyfr/plugins/integrate.py:
```diff
@@ -143,24 +151,10 @@ def _eval_exprs(self, intg):
-
-                    for dim, grad in zip('xyz', gradpn):
+            # Prepare any required gradient
+            if self._gradpinfo:
```
This if statement can go.

GonzaloSaez replied:
Done

FreddieWitherden commented on pyfr/plugins/tavg.py:
```diff
@@ -118,24 +126,10 @@ def _eval_acc_exprs(self, intg):
-                    gradpn = gradpn.reshape(self.ndims, nupts, -1)
-
-                    for dim, grad in zip('xyz', gradpn):
+            # Prepare any required gradient
```
gradients

GonzaloSaez replied:
Done

FreddieWitherden commented:
I think we're pretty much there; last bit to tweak is the grad_con_to_pri function for Navier-Stokes.

GonzaloSaez commented:
> I think we're pretty much there; last bit to tweak is the grad_con_to_pri function for Navier-Stokes.

So if i understood correctly you want to use the following expression?

<img src="https://render.githubusercontent.com/render/math?math=\nabla p = \nabla E - \frac{1}{2}\left(\vec{u} \cdot \nabla (\rho \vec{u} ) %2B \rho\vec{u} \cdot \nabla \vec{u}  \right)">

FreddieWitherden commented:
> > I think we're pretty much there; last bit to tweak is the grad_con_to_pri function for Navier-Stokes.
> 
> So if i understood correctly you want to use the following expression?
> 
> ![](https://render.githubusercontent.com/render/math?math=\nabla p = \nabla E - \frac{1}{2}\left(\vec{u} \cdot \nabla (\rho \vec{u} ) %2B \rho\vec{u} \cdot \nabla \vec{u} \right))

I think we can simply the implementation of that expression by taking advantage of various NumPy tricks.

GonzaloSaez commented:
> > > I think we're pretty much there; last bit to tweak is the grad_con_to_pri function for Navier-Stokes.
> > 
> > 
> > So if i understood correctly you want to use the following expression?
> > ![]([https://render.githubusercontent.com/render/math?math=\nabla](https://render.githubusercontent.com/render/math?math=%5Cnabla) p = \nabla E - \frac{1}{2}\left(\vec{u} \cdot \nabla (\rho \vec{u} ) %2B \rho\vec{u} \cdot \nabla \vec{u} \right))
> 
> I think we can simply the implementation of that expression by taking advantage of various NumPy tricks.

Yes sure, I'll work a bit with `np.einsum` to implement it

GonzaloSaez commented:
I have implemented the computation of `grad_p` using np.einsum, let me know if that works for you. The previous method computation was buggy. The new one should be OK

FreddieWitherden commented on pyfr/solvers/navstokes/elements.py:
```diff
@@ -8,6 +10,28 @@ class NavierStokesElements(BaseFluidElements, BaseAdvectionDiffusionElements):
 
+    @staticmethod
+    def grad_con_to_pri(cons, grad_cons, cfg):
+        rho = cons[0]
```
rho, *rhouvw = cons[:-1]

FreddieWitherden commented on pyfr/solvers/navstokes/elements.py:
```diff
@@ -8,6 +10,28 @@ class NavierStokesElements(BaseFluidElements, BaseAdvectionDiffusionElements):
+        # Pressure gradient
+        # ∇p = (gamma - 1)*[∇E - 1/2*(\vec{u}·∇(\rho\vec{u}) - \rho\vec{u}·∇(\vec{u}))]
+        gamma = cfg.getfloat('constants', 'gamma')
+        grad_p = grad_E - 0.5*(np.einsum('ijk, iljk -> ljk', vs, grad_rhouvw) \
```
No need for \, see if the + will fit on the line above.

GonzaloSaez replied:
Not sure if I understood it correctly, let me know if the new version works for you

FreddieWitherden commented on pyfr/solvers/navstokes/elements.py:
```diff
@@ -8,6 +10,28 @@ class NavierStokesElements(BaseFluidElements, BaseAdvectionDiffusionElements):
+        grad_v = [(grad_rhov - v*grad_rho)/rho for grad_rhov, v in zip(grad_cons[1:-1], vs)]
+
+        # Pressure gradient
+        # ∇p = (gamma - 1)*[∇E - 1/2*(\vec{u}·∇(\rho\vec{u}) - \rho\vec{u}·∇(\vec{u}))]
```
Can use actual rho.

FreddieWitherden commented on pyfr/solvers/navstokes/elements.py:
```diff
@@ -8,6 +10,28 @@ class NavierStokesElements(BaseFluidElements, BaseAdvectionDiffusionElements):
+        vs = [rhov/rho for rhov in cons[1:-1]]
+
+        # Velocity gradients
+        # ∇(\vec{u}) = 1/\rho*[∇(\rho\vec{u}) - \vec{u} \otimes ∇(\rho)]
```
Can use actual ρ and ·

FreddieWitherden commented on pyfr/solvers/navstokes/elements.py:
```diff
@@ -8,6 +10,28 @@ class NavierStokesElements(BaseFluidElements, BaseAdvectionDiffusionElements):
+        grad_rho, *grad_rhouvw, grad_E = grad_cons
+        
+        # Divide momentum components by rho
+        vs = [rhov/rho for rhov in cons[1:-1]]
```
Call uvw for consistency?  Also can use rhouvw.

FreddieWitherden commented on pyfr/solvers/navstokes/elements.py:
```diff
@@ -8,6 +10,28 @@ class NavierStokesElements(BaseFluidElements, BaseAdvectionDiffusionElements):
+
+        # Velocity gradients
+        # ∇(\vec{u}) = 1/\rho*[∇(\rho\vec{u}) - \vec{u} \otimes ∇(\rho)]
+        grad_v = [(grad_rhov - v*grad_rho)/rho for grad_rhov, v in zip(grad_cons[1:-1], vs)]
```
Use grad_rhouvw?

FreddieWitherden commented on pyfr/solvers/navstokes/elements.py:
```diff
@@ -8,6 +10,27 @@ class NavierStokesElements(BaseFluidElements, BaseAdvectionDiffusionElements):
+
+        # Velocity gradients
+        # ∇(\vec{u}) = 1/ρ·[∇(ρ\vec{u}) - \vec{u} \otimes ∇ρ]
+        grad_uvw = [(grad_rhov - v*grad_rho)/rho for grad_rhov, v in zip(grad_rhouvw, uvw)]
```
Line is a tad too long

GonzaloSaez replied:
Changed

FreddieWitherden commented on pyfr/solvers/navstokes/elements.py:
```diff
@@ -8,6 +10,27 @@ class NavierStokesElements(BaseFluidElements, BaseAdvectionDiffusionElements):
+        # ∇p = (gamma - 1)·[∇E - 1/2*(\vec{u}·∇(ρ\vec{u}) - ρ\vec{u}·∇(\vec{u}))]
+        gamma = cfg.getfloat('constants', 'gamma')
+        grad_p = grad_E - 0.5*(np.einsum('ijk, iljk -> ljk', uvw, grad_rhouvw) +
+                              np.einsum('ijk, iljk -> ljk', rhouvw, grad_uvw))
```
Add space before np.

FreddieWitherden commented:
Merged!  Thank you very much for this; it is greatly appreciated.
